{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This is the notebook I used to build up the libary's core funcion.\n",
    "#### I also used it to read the gff3 file from genecode project to create the feature dataframe and save as parquet file in the resources folder.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from io import StringIO\n",
    "from os import PathLike\n",
    "from pathlib import Path\n",
    "import gzip\n",
    "import logging\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def read_gff3_to_df(feature_file_path: PathLike) -> pd.DataFrame:\n",
    "    # Define the columns in the GFF3 file\n",
    "    GFF3_columns=[\"Sequence ID\",\"source\",\"Feature Type\",\"Feature Start\",\"Feature End\",\"Score\",\"Strand\",\"Phase\",\"INFO\"]\n",
    "    \n",
    "    # Check if the file is gzipped or not\n",
    "    if feature_file_path.suffix==\".gz\":\n",
    "        FILE=gzip.open(feature_file_path, 'r')\n",
    "    elif feature_file_path.suffix==\".gff3\" or feature_file_path.suffix==\".gff\":\n",
    "        FILE=open(feature_file_path, 'r')\n",
    "    else:\n",
    "        raise ValueError(f\"The file {feature_file_path} is not a recognized file type. Please provide a .gz, .gff3 or .gff file.\")\n",
    "        \n",
    "    with FILE as file:\n",
    "        outlist=list()\n",
    "        for line in file.readlines():\n",
    "            line=line.decode().strip()\n",
    "            if not line.startswith(\"#\"):\n",
    "                # Parse the line and create a dictionary\n",
    "                parsed_dict=dict(zip(GFF3_columns,line.split(\"\\t\")))\n",
    "                # Parse the INFO field and create a dictionary\n",
    "                INFO_dict=dict([item.split(\"=\") for item in parsed_dict[\"INFO\"].split(\";\")])\n",
    "                # Update the parsed dictionary with the INFO dictionary\n",
    "                parsed_dict.update(INFO_dict)\n",
    "                parsed_dict.pop(\"INFO\")\n",
    "                outlist.append(parsed_dict)\n",
    "\n",
    "    return pd.DataFrame(outlist)\n",
    "\n",
    "# Define the file paths for GRCh38 and GRCh37 GFF3 files\n",
    "gencode_Grch38_path = Path(\"resources/gencode.v45.basic.annotation.gff3.gz\")\n",
    "gencode_Grch37_path = Path(\"resources/gencode.v45lift37.basic.annotation.gff3.gz\")\n",
    "\n",
    "# Read the GRCh38 GFF3 file into a DataFrame\n",
    "grch38_df = read_gff3_to_df(gencode_Grch38_path)\n",
    "\n",
    "# Read the GRCh37 GFF3 file into a DataFrame\n",
    "grch37_df = read_gff3_to_df(gencode_Grch37_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the cell used to create the CDS DataFrame and save it to a parquet file.\n",
    "cds_grch38_df=grch38_df[grch38_df[\"Feature Type\"]==\"CDS\"].copy()\n",
    "cds_grch38_df[\"Genome_Build\"]=\"GRCh38\" \n",
    "\n",
    "\n",
    "cds_grch37_df=grch37_df[grch37_df[\"Feature Type\"]==\"CDS\"].copy()\n",
    "cds_grch37_df[\"Genome_Build\"]=\"GRCh37\"\n",
    "\n",
    "cds_dfs=pd.concat([cds_grch38_df,cds_grch37_df])\n",
    "\n",
    "cds_dfs.reset_index(drop=True, inplace=True)\n",
    "\n",
    "cds_dfs.to_parquet(\"resources/genecode_CDS_DataFrame.parquet\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence ID</th>\n",
       "      <th>source</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>Feature Start</th>\n",
       "      <th>Feature End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>artif_dupl</th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_original_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65565</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69037</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000641515.2</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000426406.4</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000332831.5</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>924432</td>\n",
       "      <td>924948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000616016.5</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311174</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000361335.1</td>\n",
       "      <td>ENSG00000212907.2_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrM:+:10470-10766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311175</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000361381.2</td>\n",
       "      <td>ENSG00000198886.2_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrM:+:10760-12137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311176</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000361567.2</td>\n",
       "      <td>ENSG00000198786.2_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrM:+:12337-14148</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311177</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000361681.2</td>\n",
       "      <td>ENSG00000198695.2_4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrM:-:14149-14673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311178</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000361789.2</td>\n",
       "      <td>ENSG00000198727.2_2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chrM:+:14747-15887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1311179 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sequence ID   source Feature Type Feature Start Feature End Score  \\\n",
       "0              chr1   HAVANA          CDS         65565       65573     .   \n",
       "1              chr1   HAVANA          CDS         69037       70008     .   \n",
       "2              chr1   HAVANA          CDS        450740      451678     .   \n",
       "3              chr1   HAVANA          CDS        685716      686654     .   \n",
       "4              chr1   HAVANA          CDS        924432      924948     .   \n",
       "...             ...      ...          ...           ...         ...   ...   \n",
       "1311174        chrM  ENSEMBL          CDS         10470       10766     .   \n",
       "1311175        chrM  ENSEMBL          CDS         10760       12137     .   \n",
       "1311176        chrM  ENSEMBL          CDS         12337       14148     .   \n",
       "1311177        chrM  ENSEMBL          CDS         14149       14673     .   \n",
       "1311178        chrM  ENSEMBL          CDS         14747       15887     .   \n",
       "\n",
       "        Strand Phase                     ID              gene_id  ...  \\\n",
       "0            +     0  CDS:ENST00000641515.2    ENSG00000186092.7  ...   \n",
       "1            +     0  CDS:ENST00000641515.2    ENSG00000186092.7  ...   \n",
       "2            -     0  CDS:ENST00000426406.4    ENSG00000284733.2  ...   \n",
       "3            -     0  CDS:ENST00000332831.5    ENSG00000284662.2  ...   \n",
       "4            +     0  CDS:ENST00000616016.5   ENSG00000187634.13  ...   \n",
       "...        ...   ...                    ...                  ...  ...   \n",
       "1311174      +     0  CDS:ENST00000361335.1  ENSG00000212907.2_2  ...   \n",
       "1311175      +     0  CDS:ENST00000361381.2  ENSG00000198886.2_2  ...   \n",
       "1311176      +     0  CDS:ENST00000361567.2  ENSG00000198786.2_2  ...   \n",
       "1311177      -     0  CDS:ENST00000361681.2  ENSG00000198695.2_4  ...   \n",
       "1311178      +     0  CDS:ENST00000361789.2  ENSG00000198727.2_2  ...   \n",
       "\n",
       "        artif_dupl Genome_Build remap_status remap_num_mappings  \\\n",
       "0              NaN       GRCh38          NaN                NaN   \n",
       "1              NaN       GRCh38          NaN                NaN   \n",
       "2              NaN       GRCh38          NaN                NaN   \n",
       "3              NaN       GRCh38          NaN                NaN   \n",
       "4              NaN       GRCh38          NaN                NaN   \n",
       "...            ...          ...          ...                ...   \n",
       "1311174        NaN       GRCh37  full_contig                NaN   \n",
       "1311175        NaN       GRCh37  full_contig                NaN   \n",
       "1311176        NaN       GRCh37  full_contig                NaN   \n",
       "1311177        NaN       GRCh37  full_contig                NaN   \n",
       "1311178        NaN       GRCh37  full_contig                NaN   \n",
       "\n",
       "        remap_target_status remap_original_location gene_status  \\\n",
       "0                       NaN                     NaN         NaN   \n",
       "1                       NaN                     NaN         NaN   \n",
       "2                       NaN                     NaN         NaN   \n",
       "3                       NaN                     NaN         NaN   \n",
       "4                       NaN                     NaN         NaN   \n",
       "...                     ...                     ...         ...   \n",
       "1311174                 NaN      chrM:+:10470-10766         NaN   \n",
       "1311175                 NaN      chrM:+:10760-12137         NaN   \n",
       "1311176                 NaN      chrM:+:12337-14148         NaN   \n",
       "1311177                 NaN      chrM:-:14149-14673         NaN   \n",
       "1311178                 NaN      chrM:+:14747-15887         NaN   \n",
       "\n",
       "        remap_substituted_missing_target transcript_status remap_original_id  \n",
       "0                                    NaN               NaN               NaN  \n",
       "1                                    NaN               NaN               NaN  \n",
       "2                                    NaN               NaN               NaN  \n",
       "3                                    NaN               NaN               NaN  \n",
       "4                                    NaN               NaN               NaN  \n",
       "...                                  ...               ...               ...  \n",
       "1311174                              NaN               NaN               NaN  \n",
       "1311175                              NaN               NaN               NaN  \n",
       "1311176                              NaN               NaN               NaN  \n",
       "1311177                              NaN               NaN               NaN  \n",
       "1311178                              NaN               NaN               NaN  \n",
       "\n",
       "[1311179 rows x 37 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_ProteinLoc_to_GenomicLoc(CDS_df: pd.DataFrame, transcript_id: str, protein_start: int, protein_end: int, genome_build: str) -> list:\n",
    "    if genome_build == \"hg19\":\n",
    "        genome_build = \"GRCh37\"\n",
    "    elif genome_build == \"hg38\":\n",
    "        genome_build = \"GRCh38\"\n",
    "        \n",
    "    transcript_df=CDS_df[CDS_df[\"transcript_id\"].str.startswith(transcript_id) & (CDS_df[\"Genome_Build\"]== genome_build)].copy()\n",
    "\n",
    "    if transcript_df.empty:\n",
    "        logging.error(\"No transcript found\")\n",
    "        return []\n",
    "    elif transcript_df.iloc[0][\"Strand\"]==\"+\":\n",
    "        transcript_df.sort_values(\"Feature Start\",inplace=True)\n",
    "    elif transcript_df.iloc[0][\"Strand\"]==\"-\":\n",
    "        transcript_df.sort_values(\"Feature Start\",inplace=True,ascending=False)\n",
    "\n",
    "    transcript_df[\"length\"]=transcript_df[\"Feature End\"].astype(int)-transcript_df[\"Feature Start\"].astype(int)+1\n",
    "    transcript_df['CDS_start'] = (transcript_df.groupby('protein_id')['length'].cumsum() - transcript_df['length'])\n",
    "    transcript_df['CDS_end'] = transcript_df.groupby('protein_id')['length'].cumsum()\n",
    "    \n",
    "    CDS_start, CDS_end = protein_start*3, protein_end*3\n",
    "\n",
    "    overlap_CDS = transcript_df[(transcript_df[\"CDS_end\"] >= CDS_start) & (transcript_df[\"CDS_start\"] <= CDS_end)]\n",
    "\n",
    "    if overlap_CDS.empty:\n",
    "        logging.error(\"No overlap found\")\n",
    "        return[]\n",
    "    elif overlap_CDS.iloc[0][\"Strand\"] == \"-\":\n",
    "        chromosome = overlap_CDS.iloc[0][\"Sequence ID\"]\n",
    "        genomic_start = int(overlap_CDS.iloc[-1][\"Feature End\"]) - (CDS_end - overlap_CDS.iloc[-1][\"CDS_start\"])+1\n",
    "        genomic_end = int(overlap_CDS.iloc[0][\"Feature End\"]) - (CDS_start - overlap_CDS.iloc[0][\"CDS_start\"])+3\n",
    "    elif overlap_CDS.iloc[0][\"Strand\"] == \"+\":\n",
    "        chromosome = overlap_CDS.iloc[0][\"Sequence ID\"]\n",
    "        genomic_start = int(overlap_CDS.iloc[0][\"Feature Start\"]) + (CDS_start - overlap_CDS.iloc[0][\"CDS_start\"])-3\n",
    "        genomic_end = int(overlap_CDS.iloc[-1][\"Feature Start\"]) + (CDS_end - overlap_CDS.iloc[-1][\"CDS_start\"])-1\n",
    "    else:\n",
    "        logging.error(\"Format Error No strand info found in fearture file\")\n",
    "        return []\n",
    "    \n",
    "    return (chromosome,genomic_start,genomic_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_input = StringIO(\"\"\"Gene Symbol\tTranscript ID\tGenome Build\tProtein Domains\n",
    "CIC\tENST00000575354\tGRCh37\tHigh mobility groupbox domain: 199—269\n",
    "CIC\tENST00000681038\tGRCh38\tHigh mobility groupbox domain : 199-269\n",
    "BRCA1\tENST00000471181\tGRCh37\tZincfinger, RING—type: 24—658;BRCA1 , serine—richdomain: 345—507\n",
    "BRCA1\tENST00000357654\tGRCh38\tZincfinger, RING—type: 24—658;BRCA1 , serine—richdomain: 345—507\"\"\")\n",
    "protein_data = pd.read_csv(test_input, sep='\\t')\n",
    "\n",
    "output_list=list()\n",
    "for index, row in protein_data.iterrows():\n",
    "    for item in row[\"Protein Domains\"].split(\";\"):\n",
    "        \n",
    "        domain_dict=dict()\n",
    "        domain_dict[\"Domain Name\"],domain_dict[\"protein_pos\"]=item.split(\":\")\n",
    "        domain_dict[\"protein_start\"],domain_dict[\"protein_end\"]=list(domain_dict[\"protein_pos\"].replace(\"—\",\"-\").strip().split(\"-\"))\n",
    "        domain_dict.pop(\"protein_pos\")\n",
    "\n",
    "        genomic_loc=convert_ProteinLoc_to_GenomicLoc(cds_dfs,transcript_id=row[\"Transcript ID\"],protein_start=int(domain_dict[\"protein_start\"]),protein_end=int(domain_dict[\"protein_end\"]),genome_build=row[\"Genome Build\"])\n",
    "        \n",
    "        domain_dict[\"Gene Symbol\"]=row[\"Gene Symbol\"]\n",
    "        domain_dict[\"Genome_Build\"]=row[\"Genome Build\"]\n",
    "        domain_dict[\"Chrom\"]=genomic_loc[0]\n",
    "        domain_dict[\"Domain Coordinates\"]=domain_dict[\"protein_start\"]+\"-\"+domain_dict[\"protein_end\"]\n",
    "        domain_dict[\"AA Length\"]=int(domain_dict[\"protein_end\"])-int(domain_dict[\"protein_start\"])+1\n",
    "        domain_dict[\"Genomic Coordiantes\"]=f\"{genomic_loc[1]}-{genomic_loc[2]}\"\n",
    "        domain_dict[\"NUC Length\"]=int(genomic_loc[2])-int(genomic_loc[1])+1\n",
    "        \n",
    "        output_list.append(domain_dict)\n",
    "    \n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>Chrom</th>\n",
       "      <th>Domain Name</th>\n",
       "      <th>Domain Coordinates</th>\n",
       "      <th>AA Length</th>\n",
       "      <th>Genomic Coordiantes</th>\n",
       "      <th>NUC Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CIC</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>chr19</td>\n",
       "      <td>High mobility groupbox domain</td>\n",
       "      <td>199-269</td>\n",
       "      <td>71</td>\n",
       "      <td>42791709-42792003</td>\n",
       "      <td>295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CIC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr19</td>\n",
       "      <td>High mobility groupbox domain</td>\n",
       "      <td>199-269</td>\n",
       "      <td>71</td>\n",
       "      <td>42272378-42272590</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>chr17</td>\n",
       "      <td>Zincfinger, RING—type</td>\n",
       "      <td>24-658</td>\n",
       "      <td>635</td>\n",
       "      <td>41245574-41276044</td>\n",
       "      <td>30471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>chr17</td>\n",
       "      <td>BRCA1 , serine—richdomain</td>\n",
       "      <td>345-507</td>\n",
       "      <td>163</td>\n",
       "      <td>41246027-41246515</td>\n",
       "      <td>489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr17</td>\n",
       "      <td>Zincfinger, RING—type</td>\n",
       "      <td>24-658</td>\n",
       "      <td>635</td>\n",
       "      <td>43093557-43124027</td>\n",
       "      <td>30471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr17</td>\n",
       "      <td>BRCA1 , serine—richdomain</td>\n",
       "      <td>345-507</td>\n",
       "      <td>163</td>\n",
       "      <td>43094010-43094498</td>\n",
       "      <td>489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol Genome_Build  Chrom                     Domain Name  \\\n",
       "0         CIC       GRCh37  chr19   High mobility groupbox domain   \n",
       "1         CIC       GRCh38  chr19  High mobility groupbox domain    \n",
       "2       BRCA1       GRCh37  chr17           Zincfinger, RING—type   \n",
       "3       BRCA1       GRCh37  chr17       BRCA1 , serine—richdomain   \n",
       "4       BRCA1       GRCh38  chr17           Zincfinger, RING—type   \n",
       "5       BRCA1       GRCh38  chr17       BRCA1 , serine—richdomain   \n",
       "\n",
       "  Domain Coordinates  AA Length Genomic Coordiantes  NUC Length  \n",
       "0            199-269         71   42791709-42792003         295  \n",
       "1            199-269         71   42272378-42272590         213  \n",
       "2             24-658        635   41245574-41276044       30471  \n",
       "3            345-507        163   41246027-41246515         489  \n",
       "4             24-658        635   43093557-43124027       30471  \n",
       "5            345-507        163   43094010-43094498         489  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(output_list)[[\"Gene Symbol\", \"Genome_Build\", \"Chrom\", \"Domain Name\", \"Domain Coordinates\", \"AA Length\", \"Genomic Coordiantes\", \"NUC Length\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence ID</th>\n",
       "      <th>source</th>\n",
       "      <th>Feature Type</th>\n",
       "      <th>Feature Start</th>\n",
       "      <th>Feature End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>artif_dupl</th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>remap_status</th>\n",
       "      <th>remap_num_mappings</th>\n",
       "      <th>remap_target_status</th>\n",
       "      <th>remap_original_location</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>remap_substituted_missing_target</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>remap_original_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1236886</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42788857</td>\n",
       "      <td>42788923</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42284705-42284771</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236887</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42790923</td>\n",
       "      <td>42791072</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42286771-42286920</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236888</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42791158</td>\n",
       "      <td>42791392</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42287006-42287240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236889</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42791472</td>\n",
       "      <td>42791601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42287320-42287449</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236890</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42791697</td>\n",
       "      <td>42791879</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42287545-42287727</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236891</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42791962</td>\n",
       "      <td>42792127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42287810-42287975</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236892</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42793040</td>\n",
       "      <td>42793242</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42288888-42289090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236893</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42793333</td>\n",
       "      <td>42793558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42289181-42289406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236894</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42794000</td>\n",
       "      <td>42794103</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42289848-42289951</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236895</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42794385</td>\n",
       "      <td>42795618</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42290233-42291466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236896</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42795710</td>\n",
       "      <td>42795897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42291558-42291745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236897</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42796238</td>\n",
       "      <td>42796359</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42292086-42292207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236898</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42796452</td>\n",
       "      <td>42796618</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42292300-42292466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236899</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42796718</td>\n",
       "      <td>42797011</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42292566-42292859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236900</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42797108</td>\n",
       "      <td>42797433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42292956-42293281</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236901</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42797744</td>\n",
       "      <td>42797988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42293592-42293836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236902</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42798087</td>\n",
       "      <td>42798241</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42293935-42294089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236903</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42798325</td>\n",
       "      <td>42798456</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42294173-42294304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236904</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42798756</td>\n",
       "      <td>42798887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42294604-42294735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236905</th>\n",
       "      <td>chr19</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>42798976</td>\n",
       "      <td>42799343</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>CDS:ENST00000575354.6</td>\n",
       "      <td>ENSG00000079432.9_16</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>full_contig</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr19:+:42294824-42295191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sequence ID  source Feature Type Feature Start Feature End Score  \\\n",
       "1236886       chr19  HAVANA          CDS      42788857    42788923     .   \n",
       "1236887       chr19  HAVANA          CDS      42790923    42791072     .   \n",
       "1236888       chr19  HAVANA          CDS      42791158    42791392     .   \n",
       "1236889       chr19  HAVANA          CDS      42791472    42791601     .   \n",
       "1236890       chr19  HAVANA          CDS      42791697    42791879     .   \n",
       "1236891       chr19  HAVANA          CDS      42791962    42792127     .   \n",
       "1236892       chr19  HAVANA          CDS      42793040    42793242     .   \n",
       "1236893       chr19  HAVANA          CDS      42793333    42793558     .   \n",
       "1236894       chr19  HAVANA          CDS      42794000    42794103     .   \n",
       "1236895       chr19  HAVANA          CDS      42794385    42795618     .   \n",
       "1236896       chr19  HAVANA          CDS      42795710    42795897     .   \n",
       "1236897       chr19  HAVANA          CDS      42796238    42796359     .   \n",
       "1236898       chr19  HAVANA          CDS      42796452    42796618     .   \n",
       "1236899       chr19  HAVANA          CDS      42796718    42797011     .   \n",
       "1236900       chr19  HAVANA          CDS      42797108    42797433     .   \n",
       "1236901       chr19  HAVANA          CDS      42797744    42797988     .   \n",
       "1236902       chr19  HAVANA          CDS      42798087    42798241     .   \n",
       "1236903       chr19  HAVANA          CDS      42798325    42798456     .   \n",
       "1236904       chr19  HAVANA          CDS      42798756    42798887     .   \n",
       "1236905       chr19  HAVANA          CDS      42798976    42799343     .   \n",
       "\n",
       "        Strand Phase                     ID               gene_id  ...  \\\n",
       "1236886      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236887      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236888      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236889      +     1  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236890      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236891      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236892      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236893      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236894      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236895      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236896      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236897      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236898      +     1  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236899      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236900      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236901      +     0  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236902      +     1  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236903      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236904      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "1236905      +     2  CDS:ENST00000575354.6  ENSG00000079432.9_16  ...   \n",
       "\n",
       "        artif_dupl Genome_Build remap_status remap_num_mappings  \\\n",
       "1236886        NaN       GRCh37  full_contig                NaN   \n",
       "1236887        NaN       GRCh37  full_contig                NaN   \n",
       "1236888        NaN       GRCh37  full_contig                NaN   \n",
       "1236889        NaN       GRCh37  full_contig                NaN   \n",
       "1236890        NaN       GRCh37  full_contig                NaN   \n",
       "1236891        NaN       GRCh37  full_contig                NaN   \n",
       "1236892        NaN       GRCh37  full_contig                NaN   \n",
       "1236893        NaN       GRCh37  full_contig                NaN   \n",
       "1236894        NaN       GRCh37  full_contig                NaN   \n",
       "1236895        NaN       GRCh37  full_contig                NaN   \n",
       "1236896        NaN       GRCh37  full_contig                NaN   \n",
       "1236897        NaN       GRCh37  full_contig                NaN   \n",
       "1236898        NaN       GRCh37  full_contig                NaN   \n",
       "1236899        NaN       GRCh37  full_contig                NaN   \n",
       "1236900        NaN       GRCh37  full_contig                NaN   \n",
       "1236901        NaN       GRCh37  full_contig                NaN   \n",
       "1236902        NaN       GRCh37  full_contig                NaN   \n",
       "1236903        NaN       GRCh37  full_contig                NaN   \n",
       "1236904        NaN       GRCh37  full_contig                NaN   \n",
       "1236905        NaN       GRCh37  full_contig                NaN   \n",
       "\n",
       "        remap_target_status    remap_original_location gene_status  \\\n",
       "1236886                 NaN  chr19:+:42284705-42284771         NaN   \n",
       "1236887                 NaN  chr19:+:42286771-42286920         NaN   \n",
       "1236888                 NaN  chr19:+:42287006-42287240         NaN   \n",
       "1236889                 NaN  chr19:+:42287320-42287449         NaN   \n",
       "1236890                 NaN  chr19:+:42287545-42287727         NaN   \n",
       "1236891                 NaN  chr19:+:42287810-42287975         NaN   \n",
       "1236892                 NaN  chr19:+:42288888-42289090         NaN   \n",
       "1236893                 NaN  chr19:+:42289181-42289406         NaN   \n",
       "1236894                 NaN  chr19:+:42289848-42289951         NaN   \n",
       "1236895                 NaN  chr19:+:42290233-42291466         NaN   \n",
       "1236896                 NaN  chr19:+:42291558-42291745         NaN   \n",
       "1236897                 NaN  chr19:+:42292086-42292207         NaN   \n",
       "1236898                 NaN  chr19:+:42292300-42292466         NaN   \n",
       "1236899                 NaN  chr19:+:42292566-42292859         NaN   \n",
       "1236900                 NaN  chr19:+:42292956-42293281         NaN   \n",
       "1236901                 NaN  chr19:+:42293592-42293836         NaN   \n",
       "1236902                 NaN  chr19:+:42293935-42294089         NaN   \n",
       "1236903                 NaN  chr19:+:42294173-42294304         NaN   \n",
       "1236904                 NaN  chr19:+:42294604-42294735         NaN   \n",
       "1236905                 NaN  chr19:+:42294824-42295191         NaN   \n",
       "\n",
       "        remap_substituted_missing_target transcript_status remap_original_id  \n",
       "1236886                              NaN               NaN               NaN  \n",
       "1236887                              NaN               NaN               NaN  \n",
       "1236888                              NaN               NaN               NaN  \n",
       "1236889                              NaN               NaN               NaN  \n",
       "1236890                              NaN               NaN               NaN  \n",
       "1236891                              NaN               NaN               NaN  \n",
       "1236892                              NaN               NaN               NaN  \n",
       "1236893                              NaN               NaN               NaN  \n",
       "1236894                              NaN               NaN               NaN  \n",
       "1236895                              NaN               NaN               NaN  \n",
       "1236896                              NaN               NaN               NaN  \n",
       "1236897                              NaN               NaN               NaN  \n",
       "1236898                              NaN               NaN               NaN  \n",
       "1236899                              NaN               NaN               NaN  \n",
       "1236900                              NaN               NaN               NaN  \n",
       "1236901                              NaN               NaN               NaN  \n",
       "1236902                              NaN               NaN               NaN  \n",
       "1236903                              NaN               NaN               NaN  \n",
       "1236904                              NaN               NaN               NaN  \n",
       "1236905                              NaN               NaN               NaN  \n",
       "\n",
       "[20 rows x 37 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_id=\"ENST00000575354\"\n",
    "genome_build=\"GRCh37\"\n",
    "cds_dfs[cds_dfs[\"transcript_id\"].str.startswith(transcript_id) & (cds_dfs[\"Genome_Build\"]== genome_build)]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
